Generate: use `GenerationConfig` as the basis for `.generate()` parametrization

gante opened the pull request:
# What does this PR do?

This PR introduces `generation_config` as the main controller of `.generate()` calls.

In particular:
1. It adds a `from_model_config` class method to `GenerateConfig`, to load a generation config from a (legacy) model config;
2. Adds a `generation_config` argument to `.generate()`. If it is not passed, it will be loaded from a pre-determined sequence (check for `generation_config.json` -> if it fails, load from the model config);
3. Because we always have a `generation_config` in `.generate()`, which holds all parametrization, gets rid of all local variables;
4. ⚠️ Changes the arguments to `generate()` (and corresponding docstring) so as to exclude `generate_config` parameters (i.e. they were moved to `**kwargs`). This is mostly to avoid a massive docstring and list of arguments that make `.generate()` very messy at the moment -- `GenerationConfig`'s docstring explains all the ways `.generate()` can be controlled, organized by type of manipulation, while `.generate()`'s docstring focuses on the API.

Notes: I've successfully run SLOW tests of GPT2 (which has a `generate_config.json`) and BART (which does not) against this PR.

gante committed:
generate from config mvp

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_20388). All of your documentation changes will be reflected on that endpoint.

gante committed:
fix failing tests

gante committed:
max_time test

sgugger commented on src/transformers/generation/utils.py:
```diff
@@ -722,166 +724,150 @@ def _reorder_cache(self, past, beam_idx):
+                else begin_index + 1
+            )
+            if generation_config.forced_decoder_ids is not None:
+                begin_index += generation_config.forced_decoder_ids[-1][
+                    0
+                ]  # generation starts after the last token that is forced
```
Put the comment above to avoid formatting breaking the line.

sgugger reviewed (commented):
I really like the fact we start passing along a generation config instead of 100 kwargs, this feels very consistent with what happens in the model files.

I have no strong opinion on the documentation of `model.generate`. It's okay for me if we defer to `GenerationConfig` for the doc.

The main change I'd like to see is having the GenerationConfig be stored (if it exists on the Hub) when we download the model in `from_pretrained`: with the current implementation all the hub kwargs like `revision`, `token` etc are not passed along, and it doesn't feel right to have them on generate. When a user wants to use a non-standard `GenerationConfig`, they can use the `from_pretrained` method of that class and pass along those kwargs there, but for the default one, we should rely on what was passed in the call to `ModelClass.from_pretrained`.

patrickvonplaten commented:
Fully agree with @sgugger here. 

Totally ok to just link to the `GenerateConfig` doc page -> think this make the docs online also cleaner actually. 
Also I'd maybe rename `generate_config` to just `config` in generate or do you think this will cause confusion with the model's config?

patrickvonplaten commented:
Overall, this is a great improvement !

gante committed:
Load default gen config at model load time; Update docs

gante committed:
further documentation; add tests

gante committed:
adapt rag to the new structure

gante committed:
handle models not instantiated with from_pretained (like in tests)

gante committed:
better default generation config

gante committed:
add can_generate fn

gante committed:
handle legacy use case of ad hoc model config changes

gante committed:
initialize gen config from config in individual methods, if gen config is none

gante committed:
fix _get_decoder_start_token_id when called outside GenerationMixin

gante committed:
correct model config load order (set attr > model config > decoder config)

gante committed:
update rag to match latest changes

gante commented:
@sgugger @patrickvonplaten It is ready for review.

Major changes since the last review request:
1. `ModelClass.from_pretrained()` pre-loads a `generation_config` attribute to the model if a `generation_config.json` exists, as suggested above
2. Handle the case where the model config has nested dictionaries (e.g. a `decoder` component)
3. Keep full retrocompatibility, including ad hoc `model.config` changes before calling `generate` (that's why you'll see `GenerationConfig.from_model_config` in so many places, all those functions may be called independently 😓 )
4. Add documentation and enhance examples

sgugger commented on docs/source/en/main_classes/text_generation.mdx:
````diff
@@ -18,12 +18,69 @@ Each framework has a generate method for auto-regressive text generation impleme
+generation_config = GenerationConfig.from_pretrained(
+    "my_account/my_model", config_file_name="awesome_generation_config.json"
+)
+```
````
Add here a line to show how to use it `generate` maybe?

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -1058,6 +1059,15 @@ def base_model(self) -> nn.Module:
+        Returns whether this model can generate sequences with `.generate()`.
+
+        Returns:
+            :bool: Whether this model can generate sequences with `.generate()`.
```
```suggestion
            `bool`: Whether this model can generate sequences with `.generate()`.
```

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -1058,6 +1059,15 @@ def base_model(self) -> nn.Module:
+        Returns:
+            :bool: Whether this model can generate sequences with `.generate()`.
+        """
+        return hasattr(self, "prepare_inputs_for_generation")
```
Are we 100% sure all generative models will have this method in the future (and have it no)? cc @patrickvonplaten

patrickvonplaten replied:
Yes every model needs this. Actually @gante could we add a `def prepare_inputs_for_generation(self, *args, **kwargs): raise NotImplementedError` to the `GenerationMixin`?

patrickvonplaten replied:
BTW note that this check here is very similar: https://github.com/huggingface/transformers/blob/d0c1ded5f36e27cd74728c0127add5afdf1f2afa/src/transformers/generation/utils.py#L949 and probably outdated. 

We could maybe improve this check using this method here. => If a model has no `prepare_inputs_for_generation`, it cannot generate

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2356,6 +2366,28 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+                )
+                model.generation_config = generation_config
+            except EnvironmentError:
+                pass
```
Should `model.generation_config` be set to `None` here?

patrickvonplaten replied:
Actually shouldn't we always create a generation_config even if no generation_config is present yet? Or add a parameter to `from_pretrained(..., create_if_not_exist=True)` ? 

Every generation model that is created after this PR should have a generation_config.json no? So I'd say we always create it automatically from `config.json` if `create_if_not_exist=True`.

patrickvonplaten replied:
If I understand the code correctly it means that no generation config will be automatically created going forward now => so I don't see how we can make the generation config the default case this way.

patrickvonplaten replied:
I see we create it here, but then don't set it. Shouldn't we do this line: https://github.com/huggingface/transformers/pull/20388/files#r1035839081 also in the case above?

Or maybe instead of adding a `create_if_not_exist` arg we could do:
```suggestion
                logger.info("Creating generation config from model config since no generation config was found in .... Since, v4.25.0 generation configs are becoming the default way of controlling the generate method. For more information you can see: ...")
                generation_config = GenerationConfig.from_model_config(self.config)
           except EnvironmentError:
```

sgugger replied:
We could also save the `generation_config` in `from_pretrained` so that we get it automatically on the repo when pushing to the Hub, maybe?

sgugger reviewed (commented):
Awesome work! I only have a couple of comments. @patrickvonplaten if you could find some time to review this PR this week as you know the insides of `generate` better than me, that would be awesome!

patrickvonplaten commented on docs/source/en/main_classes/text_generation.mdx:
```diff
@@ -18,12 +18,69 @@ Each framework has a generate method for auto-regressive text generation impleme
 - Flax/JAX [`~generation.FlaxGenerationMixin.generate`] is implemented in [`~generation.FlaxGenerationMixin`].
 
-<!--- TODO: add a brief description of GenerationConfig (with examples) when it becomes usable with generate --->
+Regardless of your framework of choice, you can parametrize the generate method with a [`~generation.GenerationConfig`]
```
```suggestion
Regardless of your framework of choice, you can parameterize the generate method with a [`~generation.GenerationConfig`]
```

patrickvonplaten commented on docs/source/en/main_classes/text_generation.mdx:
```diff
@@ -18,12 +18,69 @@ Each framework has a generate method for auto-regressive text generation impleme
+<Tip>
+
+If you inspect a serialized [`~generation.GenerationConfig`] file or print a class instance, you will notice that
+default values are omitted. Some attributes, like `max_new_tokens`, have a conservative default value, to avoid running
```
`max_length` as well no?
```suggestion
default values are omitted. Some attributes, like `max_new_tokens` and `max_length`, have a conservative
```

patrickvonplaten replied:
`max_new_tokens` defaults to `None`. I think it's `max_length` that defaults to 20 no? 
Honestly I don't think we can never remove `max_length` from the config and also not remove the default. It's simply used too much and `max_new_tokens` is harder to serve as a good default.

So maybe just say "max_length" here?
```suggestion
default values are omitted. Some attributes, like `max_length`, have a conservative default value, to avoid running
```

(nit) Also `top_k` has a default set to 50 which could be mentioned here.

patrickvonplaten commented on docs/source/en/main_classes/text_generation.mdx:
```diff
@@ -18,12 +18,69 @@ Each framework has a generate method for auto-regressive text generation impleme
+from transformers import GenerationConfig
+
+# Create a new generation configuration for later use
+my_awesome_config = GenerationConfig(num_beams=8, do_sample=False, bad_words_ids=[[42, 43, 44]], eos_token_id=2)
```
Maybe use some "real-world" example? E.g. using T5 for both translation and QA?

patrickvonplaten commented on src/transformers/models/rag/modeling_rag.py:
```diff
@@ -1497,53 +1441,33 @@ def generate(
+            if self.generation_config is not None:
+                generation_config = self.generation_config
+            else:
+                generation_config = GenerationConfig.from_model_config(self.config)
```
I see we create the generation_config here

patrickvonplaten commented on src/transformers/generation/utils.py:
```diff
@@ -2728,6 +2548,11 @@ def beam_search(
+        generation_config = (
+            self.generation_config
+            if self.generation_config is not None
+            else GenerationConfig.from_model_config(self.config)
```
IMO the better place is to do this in `from_pretrained(...)` and `__init__(...)` of the model instead of here as mentioned in: https://github.com/huggingface/transformers/pull/20388/files#r1033599800

patrickvonplaten reviewed (commented):
Looks very nice.

Just one major thing:

I think we should generate the generation_config when doing `from_pretrained(...)` and `__init__(...)` of a model that is capable of using `generate` and then also directly save a `generate_config.json`. Otherwise `GenerationConfig.from_model_config(self.config)` is called over and over again in generate and people won't switch to using the generation config really IMO. Wdyt @sgugger @gante ?

Apart from this just left some nits.

sgugger commented:
Agreed with you @patrickvonplaten , that's a very good idea!

gante committed:
Apply suggestions from code review

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

gante committed:
load gen config from model config in model.from_pretrained

gante committed:
fix can_generate fn

gante committed:
handle generate calls without a previous from_pretrained (e.g. tests)

gante committed:
add legacy behavior (and a warning)

gante commented:
@sgugger @patrickvonplaten 

Here is a summary of the key changes since your last review:
- (thanks for the suggestion!) In `model.from_pretrained`, `model.generation_config` is set from the model config if the generation config doesn’t exist, effectively making all future generation-capable models hold a default generation config parameter. NOTE: This required minor legacy handling logic, for the case where the user makes ad hoc model config changes to control generation (which the previous solution intentionally accounted for)
- added a default `prepare_inputs_for_generation`, which raises `NotImplementedError`, and updated the new `can_generate` check accordingly. Contrarily to @patrickvonplaten's suggestion, I've kept the  `_validate_model()` check -- it returns an informative exception to the user if they try to generate with an incorrect class of a model with generation capabilities, like `AutoModel.from_pretrained(“gpt2”)`. Not using the right class was a common source of issues in the past.
- Improved the example to use named generation config files with an actual T5 example. I think two named generation configs would make the example too long 🤔 (cc @patrickvonplaten) 

I was thinking of doing the following in a follow-up PR (to avoid adding more features to this already long PR that is blocking Arthur on Whisper work):
- Add the needed modifications such that `model.save_pretrained` can push to the hub a default generation config if the file doesn’t yet exist, from the `model.generation_config` parameter (as @sgugger suggested)

sgugger commented on src/transformers/modeling_utils.py:
```diff
@@ -2477,6 +2490,29 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
+                    **kwargs,
+                )
+            except OSError:
+                logger.warning(
```
Maybe a warning is a bit strong here? An info is enough probably, since it will appear at each call to `from_pretrained`.

sgugger reviewed (approved):
Looking great, thanks!

gante committed:
lower logger severity

gante commented:
@patrickvonplaten -- I'm merging to unblock @ArthurZucker's work on Whisper. 

Comments to the points above are still helpful, and I can include them in a subsequent PR! :D

fxmarty commented:
The addition of `can_generate()` is breaking in Optimum, where we use `generate()` on models which do not inherit from `PreTrainedModel`. Why isn't `can_generate()` in `GenerationMixin`? cc @gante

gante commented:
@fxmarty `can_generate()` is called in `PreTrainedModel` at initialization time, to initialize the (new) generation config if it's a generation-compatible model. All models in `transformers` inherit `GenerationMixin`, but in fact `can_generate()` is tangling the two classes at the moment, which is undesirable.

I may be able to rework this part, but I need to know -- what breaks on your end exactly?

fxmarty commented:
> All models in transformers inherit GenerationMixin

Yes thanks, I forgot this part!

The PR I linked fix the issue on our end. I think what is breaking is that `generate()` is no more usable on models that are not inheriting from `PreTrainedModel` or that don't redefine `can_generate()`, because of https://github.com/huggingface/transformers/blob/8637316e5e94ba0a2493e5df7846f2f23f46eaef/src/transformers/generation/utils.py#L934

But it's a very minor issue, and the fix is easy, so it's probably not too important.

gante commented:
@fxmarty 👍 

In the long run, I'd like to see if it's possible to separate the two (`PreTrainedModel` and `GenerationMixin`, where a model only inherits `GenerationMixin` if it can generate). It should help libraries downstream like `optimum`!

Let me know if I can be of further assistance.
